### PR TITLE
Enabling support of rerankers models 2B and 8B of qwen3vl

### PR DESCRIPTION
**Pull request by quic-amitraj**
This PR adds support for the following Qwen3-VL reranker models on AI100:

  - `Qwen/Qwen3-VL-Reranker-2B`
  - `Qwen/Qwen3-VL-Reranker-8B`

  The support is implemented using the existing QEff image-text-to-text flow (dual QPC), with model parity validation focused on **PyTorch(original) vs AI100**.
  
  ### Results

  | Model | PyTorch score | AI100 score | MAD max | Status |
  |---|---:|---:|---:|---|
  | Qwen/Qwen3-VL-Reranker-2B | 0.3213230073 | 0.3259495199  | 4.626513e-03 | Pass |
  | Qwen/Qwen3-VL-Reranker-8B | 0.6058825254 | 0.6043989062 |  1.483619e-03 | Pass |

### CODE REVIEW

File: tests/transformers/models/reranker/test_reranker_mad.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+import torch
+from huggingface_hub import snapshot_download
+from qwen_vl_utils import process_vision_info
+from transformers import AutoConfig, AutoProcessor
+
+from QEfficient.generation.cloud_infer import QAICInferenceSession
+from QEfficient.transformers.models.modeling_auto import QEFFAutoModelForImageTextToText
+from QEfficient.utils.test_utils import load_vlm_model, set_num_layers_vlm
+
+CONFIG_PATH = "tests/configs/image_text_model_configs.json"
```
Comment (quic-rishinr) [inline tests/transformers/models/reranker/test_reranker_mad.py]: nit: better to relative path instead of hard path similar to other tests

File: tests/unit_test/models/reranker/test_reranker_models_unit.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import json
+from pathlib import Path
+from typing import Dict, List
+
+import pytest
+from transformers import AutoConfig
+
+from QEfficient.utils.test_utils import set_num_layers_vlm
+
+CONFIG_PATH = Path("tests/configs/image_text_model_configs.json")
```
Comment (quic-rishinr) [inline tests/unit_test/models/reranker/test_reranker_models_unit.py]: same as above, better to use absolute path similar to other test file like os.path.join(os.path.dirname(__file__), config_path))

File: examples/reranker/qwen3vl/reranker_model.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
+# -----------------------------------------------------------------------------
+#
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# -----------------------------------------------------------------------------
+
+"""Core AI100 reranker implementation for Qwen3-VL reranker models."""
+
+import os
```
Comment (quic-rishinr) [inline examples/reranker/qwen3vl/reranker_model.py]: Do we need this script as part of example? example file should show the api usage and it should be simple.

File: QEfficient/transformers/models/qwen3_vl/modeling_qwen3_vl.py [hunk 4]
```diff
@@ -830,8 +830,13 @@ def get_dummy_inputs(
         continuous_batching: bool = False,
         **kwargs,
     ):
+        prefill_seq_len = kwargs.get("prefill_seq_len")
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/qwen3_vl/modeling_qwen3_vl.py]: nit: Can we simplify it to prefill_seq_len = int(kwargs.get("prefill_seq_len", constants.ONNX_EXPORT_EXAMPLE_SEQ_LEN)) as done in other files

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

File: examples/reranker/qwen3vl/reranker_model.py (hunk 3 above)
Comment (quic-amitraj) [reply examples/reranker/qwen3vl/reranker_model.py]: Thanks for the suggestion. We kept this module intentionally because the Qwen3-VL reranker requires model-specific preprocessing + yes/no scoring flow from the official reference script: https://huggingface.co/Qwen/Qwen3-VL-Reranker-2B/blob/main/scripts/qwen3_vl_reranker.py.

In our PR, the user-facing example remains simple (`qwen3_vl_reranker.py`), and `reranker_model.py` contains the internal implementation needed to adapt that reference behavior to AI100 dual-QPC execution. If we remove this helper and keep only a thin API script, we lose parity with the upstream reranker logic and AI100 support path.

I will add a short note in the file header/README to make this intent explicit.

File: tests/transformers/models/reranker/test_reranker_mad.py (hunk 1 above)
Comment (quic-amitraj) [reply tests/transformers/models/reranker/test_reranker_mad.py]: Updated. I switched this config path to be file-relative using `os.path.join(os.path.dirname(__file__), "../../../configs/image_text_model_configs.json")`, matching the existing test pattern and avoiding repo-root dependent hardcoded paths.

File: tests/unit_test/models/reranker/test_reranker_models_unit.py (hunk 2 above)
Comment (quic-amitraj) [reply tests/unit_test/models/reranker/test_reranker_models_unit.py]: Updated similarly here as well. The config file is now resolved relative to the test file path using `os.path.join(os.path.dirname(__file__), "../../../configs/image_text_model_configs.json")`, consistent with other tests.

File: QEfficient/transformers/models/qwen3_vl/modeling_qwen3_vl.py (hunk 4 above)
Comment (quic-amitraj) [reply QEfficient/transformers/models/qwen3_vl/modeling_qwen3_vl.py]: Done. Simplified to:
`prefill_seq_len = int(kwargs.get("prefill_seq_len", constants.ONNX_EXPORT_EXAMPLE_SEQ_LEN))`
for consistency with the pattern used in other files.

File: examples/reranker/qwen3vl/reranker_model.py (hunk 3 above)
Comment (quic-amitraj) [reply examples/reranker/qwen3vl/reranker_model.py]: Added explicit note in both places to make this clear:
- `examples/reranker/qwen3vl/reranker_model.py` module docstring now references the official Qwen script and explains why this helper exists.
- `examples/reranker/qwen3vl/README.md` now clarifies that `qwen3_vl_reranker.py` is the simple API example and `reranker_model.py` is the AI100-adapted implementation module.

File: examples/reranker/qwen3vl/qwen3_vl_reranker.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        "fps": 1.0,
+    }
+
+
+def main() -> None:
+    """Run AI100 reranker inference and print scores."""
+    args = parse_args()
+
+    # Initialize the AI100-only reranker wrapper.
+    model = QEffQwen3VLReranker(
```
Comment (quic-rishinr) [inline examples/reranker/qwen3vl/qwen3_vl_reranker.py]: instead of wrapping everything lets expose the use of QEFFAutoModelForImageTextToText and for loading the model, compile. This way the example script will be aligned with our other examples and user will know how to generate the QPCs using Qeff APIs
Also if the generation is genetic lets keep it under QEfficient/generation

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

**Review by quic-rishinr — APPROVED** [review]

File: examples/reranker/qwen3vl/qwen3_vl_reranker.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+def main() -> None:
+    """Run AI100 reranker inference and print scores."""
+    args = parse_args()
+
+    # Resolve model source (HF repo id -> local snapshot path for stable loading).
+    model_source = resolve_model_source(args.model_name)
+
+    # 1) Load config + processor + QEff model through public QEff/HF APIs.
+    config = AutoConfig.from_pretrained(model_source, trust_remote_code=True, padding=True)
```
Comment (quic-mamta) [inline examples/reranker/qwen3vl/qwen3_vl_reranker.py]: why are we passing padding argument here, please remove.

File: QEfficient/transformers/models/whisper/modeling_whisper.py [hunk 7]
```diff
@@ -792,9 +796,13 @@ def forward(
 
     def get_dummy_inputs(
         self,
+        **kwargs,
     ):
         bs = 1
-        seq_len = 32
+        seq_len = kwargs.get("prefill_seq_len")
+        if seq_len is None:
+            seq_len = 32
```
Comment (quic-mamta) [inline QEfficient/transformers/models/whisper/modeling_whisper.py]: can we use constants.ONNX_EXPORT_EXAMPLE_SEQ_LEN here as well?

File: examples/reranker/qwen3vl/reranker_model.py [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+        non_special_kept_count = 0
+        for token in tokens:
+            if token in special_tokens_set:
+                final_tokens.append(token)
+            elif non_special_kept_count < num_non_special_to_keep:
+                final_tokens.append(token)
+                non_special_kept_count += 1
+        return final_tokens
+
+    def _format_mm_content(self, text, image, video, prefix: str) -> List[Dict]:
```
Comment (quic-mamta) [inline examples/reranker/qwen3vl/reranker_model.py:99]: Please move this and similar functions being used in tests to a common file to avoid duplication.

File: QEfficient/transformers/models/whisper/modeling_whisper.py (hunk 7 above)
Comment (quic-amitraj) [reply QEfficient/transformers/models/whisper/modeling_whisper.py]: Done. Replaced the hardcoded `seq_len = 32` block with `seq_len = int(kwargs.get("prefill_seq_len", ONNX_EXPORT_EXAMPLE_SEQ_LEN))` and added `ONNX_EXPORT_EXAMPLE_SEQ_LEN` to the import from `QEfficient.utils.constants`. Behavior is unchanged since the constant resolves to 32.

File: examples/reranker/qwen3vl/qwen3_vl_reranker.py (hunk 6 above)
Comment (quic-amitraj) [reply examples/reranker/qwen3vl/qwen3_vl_reranker.py]: Done. Removed `padding=True` from both `AutoConfig.from_pretrained` and `AutoProcessor.from_pretrained`.

File: examples/reranker/qwen3vl/reranker_model.py (hunk 8 above)
Comment (quic-amitraj) [reply examples/reranker/qwen3vl/reranker_model.py:99]: The `resolve_model_source` function in `reranker_model.py` is an example-level helper and intentionally kept local there. Tests can't cleanly import from `examples/`, so the test file defines its own minimal version of the same logic. Moving it to `QEfficient/utils/` would be the right approach for a shared utility, but given this is reranker-specific preprocessing tied to the example script, we'd prefer to keep it scoped here to avoid widening the package surface for a single model's example. Happy to revisit if this becomes a pattern across more models.

File: examples/reranker/qwen3vl/qwen3_vl_reranker.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    # 5) Run AI100 scoring on precompiled QPCs.
+    scores = reranker.process(
+        inputs=inputs,
+        qpc_paths=qpc_paths,
+        prefill_seq_len=compile_specs["prefill_seq_len"],
+    )
+
+    print(scores)
+    # [0.8624675869941711, 0.6706082820892334, 0.8116759657859802]
```
Comment (quic-hemagnih) [inline examples/reranker/qwen3vl/qwen3_vl_reranker.py:133]: nit - clean up this

File: scripts/Jenkinsfile [hunk 10]
```diff
@@ -64,6 +64,7 @@ pipeline {
                     pip install .[test] &&
                     pip install junitparser pytest-xdist &&
                     pip install librosa==0.10.2 soundfile==0.13.1 &&
+                    pip install qwen-vl-utils==0.0.14 &&
```
Comment (quic-hemagnih) [inline scripts/Jenkinsfile:67]: I think we have legals for pkg:pypi/qwen-vl-utils@0.0.8, 
Did any one raised the legals for it? else we need to raise it.

**Review by quic-hemagnih — CHANGES_REQUESTED** [review]
Code is nicely organized and well written. Please check the legals for qwen-vl-utils package for version 0.0.14

File: examples/reranker/qwen3vl/qwen3_vl_reranker.py (hunk 9 above)
Comment (quic-amitraj) [reply examples/reranker/qwen3vl/qwen3_vl_reranker.py:133]: Intentionally kept it, its the score he should get after running the example script. Following the same pattern from hf- https://huggingface.co/Qwen/Qwen3-VL-Reranker-2B(please refer the example section)

**Comment by quic-hemagnih** [comment]
Lets merge on the tmp branch, LGTM

**Review by quic-hemagnih — APPROVED** [review]

File: QEfficient/transformers/models/mistral3/modeling_mistral3.py [hunk 11]
```diff
@@ -346,8 +346,12 @@ def get_dummy_inputs(
         continuous_batching: bool = False,
         **kwargs,
     ):
+        prefill_seq_len = kwargs.get("prefill_seq_len")
+        if prefill_seq_len is None:
+            prefill_seq_len = constants.ONNX_EXPORT_EXAMPLE_SEQ_LEN
+        prefill_seq_len = int(prefill_seq_len)
```
Comment (vbaddi) [inline QEfficient/transformers/models/mistral3/modeling_mistral3.py:352]: why is this change made?
